alf/experience_replayers/replay_buffer_test.py

```diff
@@ -407,6 +409,30 @@ def test_replay_buffer(self, allow_multiprocess):
         replay_buffer.clear()
         self.assertEqual(replay_buffer.total_size, 0)
 
+    def test_without_replacement(self):
+        num_envs = 4
+        max_length = 100
+        replay_buffer = ReplayBuffer(
+            data_spec=self.data_spec,
+            num_environments=num_envs,
+            max_length=max_length,
+            with_replacement=False,
+            recent_data_ratio=0.5,
+            recent_data_steps=4)
+        replay_buffer.add_batch(get_batch([0, 1, 2, 3], self.dim, t=0, x=0.))
+        batch, info = replay_buffer.get_batch(4, 1)
+        self.assertEqual(info.env_ids, torch.tensor([0, 1, 2, 3]))
+
+        replay_buffer.add_batch(get_batch([0, 1, 2, 3], self.dim, t=1, x=1.0))
+        batch, info = replay_buffer.get_batch(8, 1)
+        self.assertEqual(info.env_ids, torch.tensor([0, 1, 2, 3] * 2))
+
+        for t in range(2, 16):
+            replay_buffer.add_batch(
+                get_batch([0, 1, 2, 3], self.dim, t=t, x=t))
+        batch, info = replay_buffer.get_batch(32, 1)
+        self.assertEqual(info.env_ids[16:], torch.tensor([0, 1, 2, 3] * 4))
+
```

Comment:
add a test that recent data is really being sampled out.  Say when sampling 32 with 4 env, all 4 recent steps data are being sampled out.

Reply:
added